### ISSUE TITLE
Refractor TOdfParagraph

### ISSUE THREAD

**Issue by joecare99** [dgaspary/fpOdf#5]
Hi, 
I would like to move the Methods from TOdfParagraph to TOdfContent, because all Content-descenders should be able to add a span, a tab, a linebeak, a nbsp ..., a hyperlink, (except of a bookmark but this doesn't really matter.
Is it OK to submit a new pull with this ?

**Comment by joecare99** [comment]
You could make a branch to test the changes before committing to master.

**Comment by dgaspary** [comment]
> You could make a branch to test the changes before committing to master.

Sure, no problem. 

I Just ask for some patience. Can take some time until I can Review.

**Comment by joecare99** [comment]
Changes are in joecare99/NewFeatures_1

**Comment by dgaspary** [comment]
I added a new example (SearchAndReplace) that shows the use of XPath Search.

XPathSearch needs some work(the accept parameter is not being considered yet) but its a good way to avoid coding Search more than needed.

At TOdfDocument.WritePackage, subprocedure RemoveUnusedStyles I use Xpath to search for Elements with the style:name attribute.

Could be a good reference for the use of XPathSearch. Although the procedure itself need revision, I suspect some problems may occur.

**Comment by dgaspary** [comment]
Hi.

I've been reviewing the code and only then I noticed the use of TFont, TFontStyles from the Graphics Unit.

That unit is part of LCLBase, and it needs to be removed from the package dependencies due to cause problems to compile in environments without graphic interface like servers. 

Maybe local copies of the TFont and TFontStyles are the easiest way to solve the problem.

Besides that, the file incs/styles-decl.inc has some font related types like TOdfFontWeight and TOdfFontStyle. 

What do you think is the best course of action?

**Comment by joecare99** [comment]
If we want TFont and TFontStyles we should use the original from graphics, you have incompatibilities otherwise.A possible solution would be to create a class TOdfFont with all the possibilities of  an OdfDocument... and a second LCL-Based package with helper-functions to convert TFont and TFontStyle to TOdfFont.

    On Monday, November 4, 2019, 9:55:13 PM GMT+1, Daniel F. Gaspary <notifications@github.com> wrote:  
 
 
Hi.

I've been reviewing the code and only then I noticed the use of TFont, TFontStyles from the Graphics Unit.

That unit is part of LCLBase, and it needs to be removed from the package dependencies due to cause problems to compile in environments without graphic interface like servers.

Maybe local copies of the TFont and TFontStyles are the easiest way to solve the problem.

Besides that, the file incs/styles-decl.inc has some font related types like TOdfFontWeight and TOdfFontStyle.

What do you think is the best course of action?

**Comment by dgaspary** [comment]
On Mon, Nov 4, 2019 at 7:15 PM Joe Care <notifications@github.com> wrote:
>
> ...A possible solution would be to create a class TOdfFont with all the possibilities of  an OdfDocument...
> and a second LCL-Based package with helper-functions to convert TFont and TFontStyle to TOdfFont.

This seems to be the most viable option.

Can you include this change with the your "NewFeatures_1" changes ?

**Comment by joecare99** [comment]
Yes, but it'll take a while ... (I started a new job yesterday, that keeps me busy a lot.)

    On Monday, November 4, 2019, 11:23:54 PM GMT+1, Daniel F. Gaspary <notifications@github.com> wrote:  
 
 On Mon, Nov 4, 2019 at 7:15 PM Joe Care <notifications@github.com> wrote:
>
> ...A possible solution would be to create a class TOdfFont with all the possibilities of an OdfDocument...
> and a second LCL-Based package with helper-functions to convert TFont and TFontStyle to TOdfFont.

This seems to be the most viable option.

Can you include this change with the your "NewFeatures_1" changes ?

**Comment by dgaspary** [comment]
On Mon, Nov 4, 2019 at 8:03 PM Joe Care <notifications@github.com> wrote:
>
> Yes, but it'll take a while ... (I started a new job yesterday, that keeps me busy a lot.)

Don't worry then. I'll make these changes and commit them.

Good luck on your new job.
